packages/firestore/src/core/sync_engine.ts

```diff
@@ -721,12 +741,52 @@ export class SyncEngine implements RemoteSyncer, SharedClientStateSyncer {
   }
 
   // PORTING NOTE: Multi-tab only
-  applyPrimaryState(isPrimary: boolean): Promise<void> {
-    this.isPrimary = isPrimary;
-    if (this.isPrimary && this.networkAllowed) {
-      return this.remoteStore.enableNetwork();
-    } else {
-      return this.remoteStore.disableNetwork();
+  async applyPrimaryState(isPrimary: boolean): Promise<void> {
+    if (isPrimary === true && this.isPrimary !== true) {
+      this.isPrimary = true;
+
+      // Secondary tabs only maintain Views for their local listeners and the
+      // Views internal state may not be 100% populated (in particular
+      // secondary tabs don't track syncedDocuments, the set of documents the
+      // server considers to be in the target). So when a secondary becomes
+      // primary, we need to need to make sure that all views for all targets
+      // match the state on disk.
+      let p = Promise.resolve();
+      const activeTargets = this.sharedClientState.getAllActiveQueryTargets();
+      activeTargets.forEach(targetId => {
+        p = p.then(async () => {
+          let queryData;
+          const query = await this.localStore.getQueryForTarget(targetId);
+          if (this.queryViewsByTarget[targetId] === undefined) {
+            // For queries that never executed on this client, we need to
+            // allocate the query in LocalStore and initialize a new View.
+            queryData = await this.localStore.allocateQuery(query);
+            await this.initializeViewAndComputeInitialSnapshot(
+              queryData,
+              false
+            );
+          } else {
+            // For queries that have a local View, we need to update their state
+            // in LocalStore (as the resume token and the snapshot version
+            // might have changed) and reconcile their views with the persisted
+            // state (the list of syncedDocuments may have gotten out of sync).
+            await this.localStore.releaseQuery(query, true);
+            queryData = await this.localStore.allocateQuery(query);
+            await this.synchronizeLocalView(queryData);
+          }
+          this.remoteStore.listen(queryData);
+        });
+      });
+      await p;
+
+      if (this.networkAllowed) {
+        await this.remoteStore.enableNetwork();
+      }
+    } else if (isPrimary === false && this.isPrimary !== false) {
```

Comment:
I think we could also get rid of the subtlety with isPrimary===undefined by just calling `remoteStore.disableNetwork()` by default (in start() I guess) and wait for applyPrimaryState() to be called before we enable it.

Reply:
As discussed earlier, this created some hurdles so I opted to leave it as is.